File: kotlin-react-router-dom-legacy/karakum-legacy/react-router/karakum/plugins/convertSkippedGenerics.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isQualifiedName(node.typeName)
        && ts.isIdentifier(node.typeName.right)
        && (
            node.typeName.right.text === "ReactElement"
            || node.typeName.right.text === "ComponentType"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Location"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "UIMatch"
        )
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*, *>`
    }

    return null
}

File: kotlin-testing-library-react/karakum/src/jsMain/kotlin/testing/library/react/karakum/plugins/convertSkippedGenerics.kt
package testing.library.react.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertSkippedGenerics = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(
            typeName.text == "RenderOptions"
                    || typeName.text == "RenderResult"
        )
        ensure(node.typeArguments == null)

        "${render(node.typeName)}<*, *>"
    }
}
